duckduckgo/macos-browser
# Encapsulate NetP subjects within observers

Task/Issue URL: https://app.asana.com/0/0/1205210093476964/f
BSK PR: https://github.com/duckduckgo/BrowserServicesKit/pull/447
CC:

**Description**:

For [Update Status View to match designs](https://app.asana.com/0/0/1205084446087078), I planned to rely on the shared Connection*Observer types to receive events from the PacketTunnnelProvider and other NetP types. However, on doing this, I noticed that we’re exposing the Subject type in order to synchronously access the most recent state.

Exposing these types as it means these events can be sent from anywhere and not just the source of the events. Rather than letting dependency on this propagate into iOS too, I decided to do a quick refactor to clean it up.

Additionally, I’ve created the associated mock types which I will also need throughout the [iOS: Network Protection MVP TestFlight build for internal testing](https://app.asana.com/0/0/1204630829332227) project. I decided now was the right time to create them as these types have been altered.

Creating the MockTunnelController is included in this BSK release though it seems somewhat independent. This is so that it can be included in the same BSK release as it will also be needed for the parent task.

**Steps to test this PR**:

**Test StatusConnectionObserver**
1. Connect to NetP
2. Disconnect from NetP
3. Ensure connection state updates correctly

**Test ConnectionServerInfoObserver**
1. Connect to NetP
2. Change the server location through the debug menu
3. Check his updates correctly in the UI

**Test ConnectionErrorObserver**
1. Connect to NetP
2. Simulate a tunnel error through the debug menu (Debug -> Network Protection -> Simulate Failure)
3. Check the status view to see if the error is shown

**Test ConnectivityIssueObserver**
_Preparation_
- Start Network Protection, make sure you're asked to allow notifications.
- Allow notifications
- Save this file to your desktop. It's a script to temporarily block UDP traffic.
- [ruleset.txt](https://github.com/duckduckgo/macos-browser/files/12018072/ruleset.txt)

_Disconnection_
- Using a terminal window type sudo pfctl -ef ~/Desktop/ruleset.txt
- Wait until NetP reports a problem and make sure a notification is shown for it.
- Wait a bit more until NetP reports through a notification that the connection was stopped.
- Once you're done restore UDP traffic by executing: sudo pfctl -d
- Test 2: Reconnected notification

_Reconnection_
- Run the steps from the previous test up to, and including step 5.
- Once the notification for the connection issues is shown, run this command in a terminal window to restore UDP traffic: sudo pfctl -d
- After a while make sure NetP reports it successfully restored the connection.

**Test ControllerErrorMesssageObserver**
1. Connect to NetP
2. Simulate a Controller error through the debug menu  (Debug -> Network Protection -> Simulate Failure)
3. Check the status view to see if the error is shown



---
###### Internal references:
[Pull Request Review Checklist](https://app.asana.com/0/1202500774821704/1203764234894239/f)
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
[Pull Request Documentation](https://app.asana.com/0/1202500774821704/1204012835277482/f)
